File: sandbox/plugins/parquet-data-format/src/main/java/org/opensearch/parquet/bridge/NativeSettings.java
/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 */

package org.opensearch.parquet.bridge;

import java.util.Collections;
import java.util.Map;

/**
 * Immutable settings passed to the native Rust writer via JNI.
 * The Rust side reads values through the getter methods.
 * All fields are nullable; the native side falls back to defaults when null.
 */
public class NativeSettings {

    private final String indexName;
    private final String compressionType;
    private final Integer compressionLevel;
    private final Long pageSizeBytes;
    private final Integer pageRowLimit;
    private final Long dictSizeBytes;
    private final Boolean bloomFilterEnabled;
    private final Double bloomFilterFpp;
    private final Long bloomFilterNdv;
    private final Long sortInMemoryThresholdBytes;
    private final Integer rowGroupMaxRows;
    private final Long rowGroupMaxBytes;
    private final Integer mergeBatchSize;
    private final Integer mergeRayonThreads;
    private final Integer mergeIoThreads;
    private final Integer mergeDeferredColumnThreshold;
    private final Map<String, String> fieldEncodings;
    private final Map<String, String> fieldCompressions;
    private final Map<String, Boolean> fieldBloomFilterEnabled;
    private final Map<String, String> typeEncodings;
    private final Map<String, String> typeCompressions;
    private final Map<String, Boolean> typeBloomFilterEnabled;
    private final Map<String, Double> typeBloomFilterFpp;
    private final Map<String, Long> typeBloomFilterNdv;

    private NativeSettings(Builder builder) {
        this.indexName = builder.indexName;
        this.compressionType = builder.compressionType;
        this.compressionLevel = builder.compressionLevel;
        this.pageSizeBytes = builder.pageSizeBytes;
        this.pageRowLimit = builder.pageRowLimit;
        this.dictSizeBytes = builder.dictSizeBytes;
        this.bloomFilterEnabled = builder.bloomFilterEnabled;
        this.bloomFilterFpp = builder.bloomFilterFpp;
        this.bloomFilterNdv = builder.bloomFilterNdv;
        this.sortInMemoryThresholdBytes = builder.sortInMemoryThresholdBytes;
        this.rowGroupMaxRows = builder.rowGroupMaxRows;
        this.rowGroupMaxBytes = builder.rowGroupMaxBytes;
        this.mergeBatchSize = builder.mergeBatchSize;
        this.mergeRayonThreads = builder.mergeRayonThreads;
        this.mergeIoThreads = builder.mergeIoThreads;
        this.mergeDeferredColumnThreshold = builder.mergeDeferredColumnThreshold;
        this.fieldEncodings = builder.fieldEncodings != null ? Collections.unmodifiableMap(builder.fieldEncodings) : Collections.emptyMap();
        this.fieldCompressions = builder.fieldCompressions != null
            ? Collections.unmodifiableMap(builder.fieldCompressions)
            : Collections.emptyMap();
        this.fieldBloomFilterEnabled = builder.fieldBloomFilterEnabled != null
            ? Collections.unmodifiableMap(builder.fieldBloomFilterEnabled)
            : Collections.emptyMap();
        this.typeEncodings = builder.typeEncodings != null ? Collections.unmodifiableMap(builder.typeEncodings) : Collections.emptyMap();
        this.typeCompressions = builder.typeCompressions != null
            ? Collections.unmodifiableMap(builder.typeCompressions)
            : Collections.emptyMap();
        this.typeBloomFilterEnabled = builder.typeBloomFilterEnabled != null
            ? Collections.unmodifiableMap(builder.typeBloomFilterEnabled)
            : Collections.emptyMap();
        this.typeBloomFilterFpp = builder.typeBloomFilterFpp != null
            ? Collections.unmodifiableMap(builder.typeBloomFilterFpp)
            : Collections.emptyMap();
        this.typeBloomFilterNdv = builder.typeBloomFilterNdv != null
            ? Collections.unmodifiableMap(builder.typeBloomFilterNdv)
            : Collections.emptyMap();
    }

    public String getIndexName() {
        return indexName;
    }

    public String getCompressionType() {
        return compressionType;
    }

    public Integer getCompressionLevel() {
        return compressionLevel;
    }

    public Long getPageSizeBytes() {
        return pageSizeBytes;
    }

    public Integer getPageRowLimit() {
        return pageRowLimit;
    }

    public Long getDictSizeBytes() {
        return dictSizeBytes;
    }

    public Boolean getBloomFilterEnabled() {
        return bloomFilterEnabled;
    }

    public Double getBloomFilterFpp() {
        return bloomFilterFpp;
    }

    public Long getBloomFilterNdv() {
        return bloomFilterNdv;
    }

    public Long getSortInMemoryThresholdBytes() {
        return sortInMemoryThresholdBytes;
    }

    public Integer getRowGroupMaxRows() {
        return rowGroupMaxRows;
    }

    public Long getRowGroupMaxBytes() {
        return rowGroupMaxBytes;
    }

    public Integer getMergeBatchSize() {
        return mergeBatchSize;
    }

    public Integer getMergeRayonThreads() {
        return mergeRayonThreads;
    }

    public Integer getMergeIoThreads() {
        return mergeIoThreads;
    }

    public Integer getMergeDeferredColumnThreshold() {
        return mergeDeferredColumnThreshold;
    }

    public Map<String, String> getFieldEncodings() {
        return fieldEncodings;
    }

    public Map<String, String> getFieldCompressions() {
        return fieldCompressions;
    }

    public Map<String, Boolean> getFieldBloomFilterEnabled() {
        return fieldBloomFilterEnabled;
    }

    public Map<String, String> getTypeEncodings() {
        return typeEncodings;
    }

    public Map<String, String> getTypeCompressions() {
        return typeCompressions;
    }

    public Map<String, Boolean> getTypeBloomFilterEnabled() {
        return typeBloomFilterEnabled;
    }

    public Map<String, Double> getTypeBloomFilterFpp() {
        return typeBloomFilterFpp;
    }

    public Map<String, Long> getTypeBloomFilterNdv() {
        return typeBloomFilterNdv;
    }

    public static Builder builder() {
        return new Builder();
    }

    public static class Builder {
        private String indexName;
        private String compressionType;
        private Integer compressionLevel;
        private Long pageSizeBytes;
        private Integer pageRowLimit;
        private Long dictSizeBytes;
        private Boolean bloomFilterEnabled;
        private Double bloomFilterFpp;
        private Long bloomFilterNdv;
        private Long sortInMemoryThresholdBytes;
        private Integer rowGroupMaxRows;
        private Long rowGroupMaxBytes;
        private Integer mergeBatchSize;
        private Integer mergeRayonThreads;
        private Integer mergeIoThreads;
        private Integer mergeDeferredColumnThreshold;
        private Map<String, String> fieldEncodings;
        private Map<String, String> fieldCompressions;
        private Map<String, Boolean> fieldBloomFilterEnabled;
        private Map<String, String> typeEncodings;
        private Map<String, String> typeCompressions;
        private Map<String, Boolean> typeBloomFilterEnabled;
        private Map<String, Double> typeBloomFilterFpp;
        private Map<String, Long> typeBloomFilterNdv;

        public Builder indexName(String v) {
            this.indexName = v;
            return this;
        }

        public Builder compressionType(String v) {
            this.compressionType = v;
            return this;
        }

        public Builder compressionLevel(Integer v) {
            this.compressionLevel = v;
            return this;
        }

        public Builder pageSizeBytes(Long v) {
            this.pageSizeBytes = v;
            return this;
        }

        public Builder pageRowLimit(Integer v) {
            this.pageRowLimit = v;
            return this;
        }

        public Builder dictSizeBytes(Long v) {
            this.dictSizeBytes = v;
            return this;
        }

        public Builder bloomFilterEnabled(Boolean v) {
            this.bloomFilterEnabled = v;
            return this;
        }

        public Builder bloomFilterFpp(Double v) {
            this.bloomFilterFpp = v;
            return this;
        }

        public Builder bloomFilterNdv(Long v) {
            this.bloomFilterNdv = v;
            return this;
        }

        public Builder sortInMemoryThresholdBytes(Long v) {
            this.sortInMemoryThresholdBytes = v;
            return this;
        }

        public Builder rowGroupMaxRows(Integer v) {
            this.rowGroupMaxRows = v;
            return this;
        }

        public Builder rowGroupMaxBytes(Long v) {
            this.rowGroupMaxBytes = v;
            return this;
        }

        public Builder mergeBatchSize(Integer v) {
            this.mergeBatchSize = v;
            return this;
        }

        public Builder mergeRayonThreads(Integer v) {
            this.mergeRayonThreads = v;
            return this;
        }

        public Builder mergeIoThreads(Integer v) {
            this.mergeIoThreads = v;
            return this;
        }

        public Builder mergeDeferredColumnThreshold(Integer v) {
            this.mergeDeferredColumnThreshold = v;
            return this;
        }

        public Builder fieldEncodings(Map<String, String> v) {
            this.fieldEncodings = v;
            return this;
        }

        public Builder fieldCompressions(Map<String, String> v) {
            this.fieldCompressions = v;
            return this;
        }

        public Builder fieldBloomFilterEnabled(Map<String, Boolean> v) {
            this.fieldBloomFilterEnabled = v;
            return this;
        }

        public Builder typeEncodings(Map<String, String> v) {
            this.typeEncodings = v;
            return this;
        }

        public Builder typeCompressions(Map<String, String> v) {
            this.typeCompressions = v;
            return this;
        }

        public Builder typeBloomFilterEnabled(Map<String, Boolean> v) {
            this.typeBloomFilterEnabled = v;
            return this;
        }

        public Builder typeBloomFilterFpp(Map<String, Double> v) {
            this.typeBloomFilterFpp = v;
            return this;
        }

        public Builder typeBloomFilterNdv(Map<String, Long> v) {
            this.typeBloomFilterNdv = v;
            return this;
        }

        public NativeSettings build() {
            return new NativeSettings(this);
        }
    }
}

File: sandbox/plugins/parquet-data-format/src/main/rust/src/native_settings.rs
/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 */

use std::collections::HashMap;

use crate::field_config::FieldConfig;

#[derive(Debug, Clone, Default)]
pub struct NativeSettings {
    pub index_name: Option<String>,
    pub compression_level: Option<i32>,
    pub compression_type: Option<String>,
    pub page_size_bytes: Option<usize>,
    pub page_row_limit: Option<usize>,
    pub dict_size_bytes: Option<usize>,
    pub field_configs: Option<HashMap<String, FieldConfig>>,
    pub custom_settings: Option<HashMap<String, String>>,
    pub bloom_filter_enabled: Option<bool>,
    pub bloom_filter_fpp: Option<f64>,
    pub bloom_filter_ndv: Option<u64>,
    pub sort_columns: Vec<String>,
    pub reverse_sorts: Vec<bool>,
    pub nulls_first: Vec<bool>,
    pub sort_in_memory_threshold_bytes: Option<u64>,
    pub sort_batch_size: Option<usize>,
    pub merge_batch_size: Option<usize>,
    pub row_group_max_rows: Option<usize>,
    pub merge_rayon_threads: Option<usize>,
    pub merge_io_threads: Option<usize>,
}

impl NativeSettings {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn get_compression_type(&self) -> &str {
        self.compression_type.as_deref().unwrap_or("LZ4_RAW")
    }

    pub fn get_compression_level(&self) -> i32 {
        self.compression_level.unwrap_or(2)
    }

    pub fn get_page_size_bytes(&self) -> usize {
        self.page_size_bytes.unwrap_or(1024 * 1024)
    }

    pub fn get_page_row_limit(&self) -> usize {
        self.page_row_limit.unwrap_or(20000)
    }

    pub fn get_dict_size_bytes(&self) -> usize {
        self.dict_size_bytes.unwrap_or(2 * 1024 * 1024)
    }

    pub fn get_bloom_filter_enabled(&self) -> bool {
        self.bloom_filter_enabled.unwrap_or(true)
    }

    pub fn get_bloom_filter_fpp(&self) -> f64 {
        self.bloom_filter_fpp.unwrap_or(0.1)
    }

    pub fn get_bloom_filter_ndv(&self) -> u64 {
        self.bloom_filter_ndv.unwrap_or(100_000)
    }

    pub fn get_field_config(&self, field_name: &str) -> Option<&FieldConfig> {
        self.field_configs.as_ref()?.get(field_name)
    }

    pub fn has_field_configs(&self) -> bool {
        self.field_configs.as_ref().map_or(false, |configs| !configs.is_empty())
    }

    pub fn get_sort_in_memory_threshold_bytes(&self) -> u64 {
        self.sort_in_memory_threshold_bytes.unwrap_or(32 * 1024 * 1024)
    }

    pub fn get_sort_batch_size(&self) -> usize {
        self.sort_batch_size.unwrap_or(8192)
    }

    pub fn get_merge_batch_size(&self) -> usize {
        self.merge_batch_size.unwrap_or(100_000)
    }

    pub fn get_row_group_max_rows(&self) -> usize {
        self.row_group_max_rows.unwrap_or(1_000_000)
    }

    pub fn get_merge_rayon_threads(&self) -> Option<usize> {
        self.merge_rayon_threads
    }

    pub fn get_merge_io_threads(&self) -> Option<usize> {
        self.merge_io_threads
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_defaults() {
        let config = NativeSettings::default();
        assert_eq!(config.get_compression_type(), "LZ4_RAW");
        assert_eq!(config.get_compression_level(), 2);
        assert_eq!(config.get_page_row_limit(), 20000);
        assert_eq!(config.get_dict_size_bytes(), 2 * 1024 * 1024);
    }

    #[test]
    fn test_struct_construction() {
        let config = NativeSettings {
            compression_type: Some("SNAPPY".to_string()),
            compression_level: Some(1),
            ..Default::default()
        };
        assert_eq!(config.get_compression_type(), "SNAPPY");
        assert_eq!(config.get_compression_level(), 1);
    }

    #[test]
    fn test_field_configs() {
        use crate::field_config::FieldConfig;
        use std::collections::HashMap;

        let mut field_configs = HashMap::new();
        field_configs.insert("timestamp".to_string(), FieldConfig {
            compression_type: Some("SNAPPY".to_string()),
            compression_level: None,
        });
        let config = NativeSettings {
            compression_type: Some("ZSTD".to_string()),
            field_configs: Some(field_configs),
            ..Default::default()
        };
        assert!(config.has_field_configs());
        let fc = config.get_field_config("timestamp").unwrap();
        assert_eq!(fc.compression_type, Some("SNAPPY".to_string()));
    }
}
